### PR TITLE
Bài tập slide Share Preferences

### PR DESCRIPTION
**Pull request by Ngduong29**
## Related Tickets
- [#Task xxxxx: Content](https://edu-redmine.sun-asterisk.vn/issues/xxxx)
## WHAT
- Bài tập slide Share Preferences
## Evidence (Screenshot or Video)
image
image
image
image
image
image
image

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/blob/master/eng/android/coding_convention_android_kotlin.md | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(Other notes)*

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -110,4 +117,4 @@
             android:exported="false" />
     </application>
 
-</manifest>
+</manifest>
```
Comment (ghost) [inline app/src/main/AndroidManifest.xml]: end line

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        super.onCreate(savedInstanceState)
+        setContentView(binding.root)
+        mShowCountTextView = binding.textCount
+        mColor = ContextCompat.getColor(this, R.color.default_background)
+        mPreferences = getSharedPreferences(sharePrefFile, MODE_PRIVATE)
+
+        mCount = mPreferences!!.getInt(COUNT_KEY,0)
+        mShowCountTextView!!.text = String.format("%s",mCount)
+        mColor = mPreferences!!.getInt(COLOR_KEY,mColor!!)
+        mShowCountTextView!!.setBackgroundColor(mColor!!)
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: không dùng assert non null

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+        binding.buttonBlackBackground.setOnClickListener{
+            val color = (it.background as ColorDrawable).color
+            mShowCountTextView!!.setBackgroundColor(color)
```
Comment (cuongvd-1287) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt:36]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            mColor = color
+        }
+        binding.buttonBlueBackground.setOnClickListener{
+            val color = (it.background as ColorDrawable).color
+            mShowCountTextView!!.setBackgroundColor(color)
```
Comment (cuongvd-1287) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt:41]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            mColor = color
+        }
+        binding.buttonRedBackground.setOnClickListener{
+            val color = (it.background as ColorDrawable).color
+            mShowCountTextView!!.setBackgroundColor(color)
```
Comment (cuongvd-1287) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt:46]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt (hunk 3 above)
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]:
```suggestion
        binding.buttonBlackBackground.setOnClickListener{
            val color = (it.background as? ColorDrawable).color
            mShowCountTextView?.setBackgroundColor(color)
```

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+        binding.buttonBlackBackground.setOnClickListener{
+            val color = (it.background as ColorDrawable).color
+            mShowCountTextView!!.setBackgroundColor(color)
+            mColor = color
+        }
+        binding.buttonBlueBackground.setOnClickListener{
+            val color = (it.background as ColorDrawable).color
+            mShowCountTextView!!.setBackgroundColor(color)
+            mColor = color
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        binding.buttonRedBackground.setOnClickListener{
+            val color = (it.background as ColorDrawable).color
+            mShowCountTextView!!.setBackgroundColor(color)
+            mColor = color
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        binding.buttonGreenBackground.setOnClickListener{
+            val color = (it.background as ColorDrawable).color
+            mShowCountTextView!!.setBackgroundColor(color)
+            mColor = color
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        binding.buttonCount.setOnClickListener {
+            mCount++
+            mShowCountTextView!!.text = String.format("%s",mCount)
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        binding.buttonReset.setOnClickListener {
+            mCount = 0
+            mShowCountTextView!!.text = String.format("%s",mCount)
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+            mColor = ContextCompat.getColor(this,R.color.default_background)
+            mShowCountTextView!!.setBackgroundColor(mColor!!)
+
+            val preferencesEditor: SharedPreferences.Editor = mPreferences!!.edit()
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putInt(COUNT_KEY,mCount)
+        mColor?.let { outState.putInt(COLOR_KEY, it) }
+    }
+
+    override fun onPause() {
+        super.onPause()
+        val preferencesEditor: SharedPreferences.Editor = mPreferences!!.edit()
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        preferencesEditor.putInt(COUNT_KEY,mCount)
+        preferencesEditor.putInt(COLOR_KEY,mColor!!)
+        preferencesEditor.apply()
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: .

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+import com.sun.android.databinding.ActivitySharePreferencesBinding
+
+class SharePreferencesActivity : AppCompatActivity() {
+    val binding by lazy { ActivitySharePreferencesBinding.inflate(layoutInflater) }
+    private var mCount: Int = 0
+    private var mColor: Int = 0
+    lateinit var mShowCountTextView: TextView
+    private val COUNT_KEY: String = "count"
+    private val COLOR_KEY: String = "color"
+    lateinit var mPreferences: SharedPreferences
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: dùng lazy và để private nha

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            mShowCountTextView.setBackgroundColor(mColor)
+
+            val preferencesEditor: SharedPreferences.Editor = mPreferences.edit()
+            preferencesEditor.clear()
+            preferencesEditor.apply()
+        }
+    }
+
+    fun changeBackGround(view: View){
+        val color = (view.background as ColorDrawable).color
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]:
```suggestion
    fun changeBackGround(view: View){
        val color = (view.background as? ColorDrawable)?.color
```

File: app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    lateinit var mShowCountTextView: TextView
+    private val COUNT_KEY: String = "count"
+    private val COLOR_KEY: String = "color"
+    lateinit var mPreferences: SharedPreferences
+    private val sharePrefFile: String = "com.example.android.hellosharedprefs"
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(binding.root)
+        mShowCountTextView = binding.textCount
```
Comment (ghost) [inline app/src/main/java/com/sun/android/sharepreferences/SharePreferencesActivity.kt]: dùng binding.textCount luôn

**Pull request merged**